wazo_websocketd/bus.py

```diff
@@ -39,205 +52,404 @@
+                continue
+
+            return
 
-    def set_msg_received_callback(self, callback):
-        # Must be called before calling "connect()". Can't be changed once connected.
-        self._msg_received_callback = callback
+    async def _notify_close(self):
+        tasks = [consumer.connection_lost() for consumer in self._consumers]
+        await asyncio.gather(*tasks, loop=self._loop)
 
-    def set_connection_lost_callback(self, callback):
-        self._connection_lost_callback = callback
+
+class _BusConnectionPool:
+    def __init__(self, connection_params, *, limit=2, loop=None):
+        self._url = 'amqp://{username}:{password}@{host}:{port}/{vhost}/'.format(
+            **connection_params._asdict()
+        )
+        self._loop = loop or asyncio.get_event_loop()
+        self._limit = limit
+        self._connections = []
+        self._iterator = None
 
     @property
-    def connected(self):
-        return self._connected
+    def _size(self):
+        return len(self._connections)
+
+    @property
+    def _round_robin(self):
+        for index in cycle(range(self._size)):
+            yield self._connections[index]
 
     async def connect(self):
-        if self._closed:
-            raise Exception('already closed')
-        if self._connected:
-            raise Exception('already connected')
+        self._connections = [
+            _BusConnection(self._url, loop=self._loop) for _ in range(self._limit)
+        ]
 
-        logger.debug('connecting to bus')
-        self._connected = True
-        try:
-            self._connection = await asynqp.connect(
-                self._host,
-                self._port,
-                self._username,
-                self._password,
-                on_connection_close=self.on_connection_closed,
-            )
-            self._channel = await self._connection.open_channel()
-            self._exchange = await self._channel.declare_exchange(
-                self._exchange_name, self._exchange_type, durable=True
-            )
-            self._queue = await self._channel.declare_queue(exclusive=True)
-            self._consumer = await self._queue.consume(
-                self._msg_received_callback, no_ack=True
+        await asyncio.gather(
+            *[connection.connect() for connection in self._connections]
+        )
+        logger.info('bus connection pool initialized with %d connections', self._size)
+
+    async def disconnect(self):
+        await asyncio.gather(
+            *[connection.disconnect() for connection in self._connections]
+        )
+        logger.info('closed bus connection pool')
+
+    def connection(self):
+        if not self._iterator:
+            self._iterator = self._round_robin
+        return next(self._iterator)
+
+
+class BusService:
+    def __init__(
+        self,
+        *,
+        username='guest',
+        password='guest',
+        host='localhost',
+        port=5672,
```

Comment:
Leave these default values in config.py to avoid any confusion with time if they are changed